core/drivers/pinctrl/pinctrl.c

```diff
@@ -0,0 +1,148 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2023, Microchip
+ */
+
+#include <assert.h>
+#include <drivers/pinctrl.h>
+#include <libfdt.h>
+#include <stdio.h>
+#include <sys/queue.h>
+#include <tee_api_defines.h>
+#include <tee_api_defines_extensions.h>
+#include <tee_api_types.h>
+#include <util.h>
+
+SLIST_HEAD(pinctrl_list, pinctrl) pinctrl_list;
+
+static const char * const pin_modes[] = {
+	[PIN_MODE_BIAS_DISABLE] = "bias-disable",
+	[PIN_MODE_BIAS_PULL_UP] = "bias-pull-up",
+	[PIN_MODE_BIAS_PULL_DOWN] = "bias-pull-down",
+};
+
+TEE_Result pinctrl_parse_dt_pin_modes(const void *fdt, int node,
+				      struct pin_modes **cfg_modes)
+{
+	unsigned int i = 0;
+	unsigned int prop_count = 0;
+	struct pin_modes *cfg_modes_ptr;
+
+	for (i = 0; i < ARRAY_SIZE(pin_modes); i++) {
+		if (!fdt_getprop(fdt, node, pin_modes[i], NULL))
+			continue;
+
+		prop_count++;
+	}
+
+	cfg_modes_ptr = calloc(1, sizeof(struct pin_modes) +
+				  prop_count * sizeof(uint32_t));
+	if (!cfg_modes_ptr)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	cfg_modes_ptr->modes_count = prop_count;
+
+	prop_count = 0;
+	for (i = 0; i < ARRAY_SIZE(pin_modes); i++) {
+		if (!fdt_getprop(fdt, node, pin_modes[i], NULL))
+			continue;
+
+		cfg_modes_ptr->modes[prop_count] = i;
+		prop_count++;
+	}
```

Comment:
I don't understand why you are storing several mode.
Each pin in a pinctrl has 1 of the 3 bias related cfg, right?

Reply:
These property are actually just a few amongst the supported ones by the different bindings. So for the moment, there is only bias supported since the atmel driver only handles that, but if you look at the bindings (https://elixir.bootlin.com/linux/latest/source/Documentation/devicetree/bindings/pinctrl/pincfg-node.yaml), there are much more that can potentially be supported later.

Reply:
ok, right.
suggestion: bitstring.h could help to ease passing boolean properties. Here with:
```c
	for (i = 0; i < PINCTRL_DT_PROPS_MAX; i++)
		if (fdt_getprop(fdt, node, pinctrl_dt_prop[i], NULL))
			bit_set(i, PINCTRL_DT_PROPS_MAX);
```
with
```
enum pinctrl_dt_prop {
	PINCTRL_DT_PROP_BIAS_DISABLE,
	PINCTRL_DT_PROP_BIAS_PULL_UP,
	PINCTRL_DT_PROP_BIAS_PULL_DOWN,
	PINCTRL_DT_PROPS_MAX
};

static const char * const pinctrl_dt_prop[] = {
	[PINCTRL_DT_PROP_BIAS_DISABLE] = "bias-disable",
	...
```

Reply:
Ok acked, i'll switch to a bitmap instead of multiple `uint32_t`